store/v2alpha1/multi/store.go

```diff
@@ -876,19 +876,29 @@ func (pr *prefixRegistry) RegisterSubstore(key string, typ types.StoreType) erro
 	return nil
 }
 
-func (tlm *traceListenMixin) AddListeners(skey types.StoreKey, listeners []types.WriteListener) {
-	key := skey.Name()
-	tlm.listeners[key] = append(tlm.listeners[key], listeners...)
+func (tlm *traceListenMixin) AddListeners(keys []types.StoreKey) {
+	listener := &types.MemoryListener{}
+	for i := range keys {
+		tlm.listeners[keys[i].Name()] = listener
+	}
 }
 
 // ListeningEnabled returns if listening is enabled for a specific KVStore
 func (tlm *traceListenMixin) ListeningEnabled(key types.StoreKey) bool {
 	if ls, has := tlm.listeners[key.Name()]; has {
-		return len(ls) != 0
+		return ls != nil
 	}
 	return false
 }
 
+func (tlm *traceListenMixin) PopStateCache() []*types.StoreKVPair {
+	var cache []*types.StoreKVPair
+	for _, ls := range tlm.listeners {
+		cache = append(cache, ls.PopStateCache()...)
+	}
+	return cache
```

Comment:
This isn't deterministic.

Reply:
fixed. also added mutex as PopStateCache destroys the accumulated state. (ref: https://github.com/cosmos/cosmos-sdk/pull/14207#discussion_r1084643517)